crates/hir-ty/src/diagnostics/match_check/usefulness.rs

```diff
@@ -311,10 +324,22 @@ impl<'a, 'p> MatchCheckCtx<'a, 'p> {
         }
     }
 
-    // Rust feature described as "Allows exhaustive pattern matching on types that contain uninhabited types."
+    // Rust's unstable feature described as "Allows exhaustive pattern matching on types that contain uninhabited types."
     pub(super) fn feature_exhaustive_patterns(&self) -> bool {
-        // FIXME see MatchCheckCtx::is_uninhabited
-        false
+        *self.feature_exhaustive_patterns.get_or_init(|| {
+            let def_map = self.db.crate_def_map(self.module.krate());
+            let root_mod = def_map.module_id(def_map.root());
+            let rood_attrs = self.db.attrs(root_mod.into());
+            let mut nightly_features = rood_attrs
+                .by_key("feature")
+                .attrs()
+                .map(|attr| attr.parse_path_comma_token_tree())
+                .flatten()
+                .flatten();
+            nightly_features.any(
+            |feat| matches!(feat.segments(), [name] if name.to_smol_str() == "exhaustive_patterns"),
+        )
+        })
```

Comment:
I think we should record the all enabled features in the `DefMap` instead of parsing them here. We will need the enabled features for completions as well once we implement honoring them there (and probably other places can benefit from knowing features as well).

Reply:
Done.